client/core/transaction_factory.go

```diff
@@ -0,0 +1,450 @@
+// 	}
+
+// 	return &types.UnbondTx{
+// 		Address: addrBytes,
+// 		Height:  int(height),
+// 	}, nil
+// }
+
+// func Rebond(addrS, heightS string) (*txs.RebondTx, error) {
+// 	if addrS == "" {
+// 		return nil, fmt.Errorf("Validator address must be given with --addr flag")
+// 	}
+
+// 	addrBytes, err := hex.DecodeString(addrS)
+// 	if err != nil {
+// 		return nil, fmt.Errorf("addr is bad hex: %v", err)
+// 	}
+
+// 	height, err := strconv.ParseInt(heightS, 10, 32)
+// 	if err != nil {
+// 		return nil, fmt.Errorf("height is misformatted: %v", err)
+// 	}
+
+// 	return &types.RebondTx{
+// 		Address: addrBytes,
+// 		Height:  int(height),
+// 	}, nil
+// }
+
+type TxResult struct {
+	BlockHash []byte // all txs get in a block
+	Hash      []byte // all txs get a hash
+
+	// only CallTx
+	Address   []byte // only for new contracts
+	Return    []byte
+	Exception string
+
+	//TODO: make Broadcast() errors more responsive so we
+	// can differentiate mempool errors from other
+}
+
+// Preserve
+func SignAndBroadcast(chainID string, nodeClient client.NodeClient, keyClient keys.KeyClient, tx txs.Tx, sign, broadcast, wait bool) (txResult *TxResult, err error) {
+	// var inputAddr []byte
+	if sign {
+		_, tx, err = signTx(keyClient, chainID, tx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if broadcast {
+		// if wait {
+		// 	var ch chan Msg
+		// 	ch, err = subscribeAndWait(tx, chainID, nodeAddr, inputAddr)
+		// 	if err != nil {
+		// 		return nil, err
+		// 	} else {
+		// 		defer func() {
+		// 			if err != nil {
+		// 				// if broadcast threw an error, just return
+		// 				return
+		// 			}
+		// 			log.WithFields(log.Fields{
+		// 				"",
+		// 				}).Debug("Waiting for tx to be committed")
+		// 			msg := <-ch
+		// 			if msg.Error != nil {
+		// 				logger.Infof("Encountered error waiting for event: %v\n", msg.Error)
+		// 				err = msg.Error
+		// 			} else {
+		// 				txResult.BlockHash = msg.BlockHash
+		// 				txResult.Return = msg.Value
+		// 				txResult.Exception = msg.Exception
+		// 			}
+		// 		}()
+		// 	}
+		// }
+		var receipt *txs.Receipt
+		receipt, err = nodeClient.Broadcast(tx)
+		if err != nil {
+			return nil, err
+		}
+		txResult = &TxResult{
+			Hash: receipt.TxHash,
+		}
+		// NOTE: [ben] is this consistent with the Ethereum protocol?  It should seem
+		// reasonable to get this returned from the chain directly.  Alternatively,
+		// the benefit is that the we don't need to trust the chain node
+		if tx_, ok := tx.(*txs.CallTx); ok {
+			if len(tx_.Address) == 0 {
+				txResult.Address = txs.NewContractAddress(tx_.Input.Address, tx_.Input.Sequence)
+			}
+		}
+	}
+	return
+}
+
+//------------------------------------------------------------------------------------
+// wait for events
+
+type Msg struct {
+	BlockHash []byte
+	Value     []byte
+	Exception string
+	Error     error
+}
+
+// func subscribeAndWait(tx txs.Tx, chainID, nodeAddr string, inputAddr []byte) (chan Msg, error) {
```

Comment:
Can we add a TODO here, perhaps reference similar functionality in: 

https://github.com/eris-ltd/eris-db/blob/develop/rpc/tendermint/test/ws_helpers.go#L137

Might be a good idea to share implementation... might not...